### PR TITLE
[lldb] Calculate process address if file address + section fails

### PR DESCRIPTION
**Pull request by augusto2112**
If the resulting address from the Address constructor that takes a file address + offset fails, fallback to calculating the process address.

rdar://128481320

### CODE REVIEW

**Comment by augusto2112** [comment]
@swift-ci test

**Comment by augusto2112** [comment]
@swift-ci test

File: lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp [hunk 1]
```diff
@@ -619,18 +619,37 @@ LLDBMemoryReader::resolveRemoteAddress(uint64_t address) const {
...
+      LLDB_LOG(log,
+               "[MemoryReader] Could not calculate processs address from file "
+               "address {0:x}, no sections in {1}",
+               file_address, object_file->GetFileSpec().GetFilename());
+      return {};
+    }
+    auto sec = sec_list->GetSectionAtIndex(0);
+    auto sec_file_address = sec->GetFileAddress();
+    auto sec_load_address = sec->GetLoadBaseAddress(&m_process.GetTarget());
+    auto delta = sec_load_address - sec_file_address;
```
Comment (adrian-prantl) [inline lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp]: Maybe add an overflow check here using one of the llvm mathextras functions?

File: lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp [hunk 2]
```diff
@@ -619,18 +619,37 @@ LLDBMemoryReader::resolveRemoteAddress(uint64_t address) const {
...
+    auto process_address = file_address + delta;
```
Comment (adrian-prantl) [inline lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp]: same here

File: lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp [hunk 3]
```diff
@@ -619,18 +619,37 @@ LLDBMemoryReader::resolveRemoteAddress(uint64_t address) const {
...
+
+    resolved = Address(process_address);
+    LLDB_LOGV(log,
+              "[MemoryReader] Could not find section with file address {0:x} "
+              "and file {1}, resolved it into process address {2:x}",
+              file_address, object_file->GetFileSpec().GetFilename(),
+              process_address);
+  } else {
```
Comment (adrian-prantl) [inline lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp]: early exit?

**Comment by kastiglione** [comment]
Is it possible to write a test for this?

File: lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp [hunk 4]
```diff
@@ -619,18 +619,54 @@ LLDBMemoryReader::resolveRemoteAddress(uint64_t address) const {
...
+               sec_load_address, sec_file_address);
+      return {};
+    }
+
+    auto virtual_address = llvm::SaturatingAdd(file_address, delta, &overflow);
+    if (overflow) {
+      LLDB_LOG(log, "[MemoryReader] file address {0:x} + delta {1:x} overflows",
+               sec_load_address, sec_file_address);
+      return {};
+    }
```
Comment (kastiglione) [inline lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp]: Is `delta` supposed to be `slide`?

**Comment by augusto2112** [comment]
> Is it possible to write a test for this?

Unfortunately I don't think that this is easily testable.

**Review by kastiglione — APPROVED** [review]

**Comment by augusto2112** [comment]
@swift-ci test

**Pull request merged**